### PR TITLE
feat(format/grit): add basic formatting for where pattern

### PR DESCRIPTION
**Pull request by branberry**
## Summary

Adds a basic formatting rule for the `where` pattern

 #2455 

## Test Plan

Snapshot test demonstrates the intended formatting behavior.

### CODE REVIEW

File: crates/biome_grit_formatter/Cargo.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ biome_rowan       = { workspace = true }
 [dev-dependencies]
 biome_configuration  = { path = "../biome_configuration" }
 biome_formatter_test = { path = "../biome_formatter_test" }
+biome_fs             = { workspace = true }
```
Comment (ematipico) [inline crates/biome_grit_formatter/Cargo.toml:23]:
```suggestion
biome_fs             = { path = "../biome_fs" }
```

This is really important

File: packages/@biomejs/wasm-nodejs/package.json
Comment (dyc3) [inline packages/@biomejs/wasm-nodejs/package.json:1]: revert the changes to this file

**Review by dyc3 — APPROVED** [review]

File: crates/biome_grit_formatter/src/context.rs [hunk 2]
```diff
@@ -62,6 +62,7 @@ pub struct GritFormatOptions {
     line_width: LineWidth,
     quote_style: QuoteStyle,
     attribute_position: AttributePosition,
+    bracket_spacing: BracketSpacing,
```
Comment (ah-yu) [inline crates/biome_grit_formatter/src/context.rs:65]: Has this setting been used? I haven't seen it. Maybe we could add it when needed.

File: crates/biome_grit_formatter/src/grit/lists/definition_list.rs [hunk 3]
```diff
@@ -5,6 +5,16 @@ pub(crate) struct FormatGritDefinitionList;
 impl FormatRule<GritDefinitionList> for FormatGritDefinitionList {
     type Context = GritFormatContext;
     fn fmt(&self, node: &GritDefinitionList, f: &mut GritFormatter) -> FormatResult<()> {
-        format_verbatim_node(node.syntax()).fmt(f)
+        let mut join = f.join_nodes_with_hardline();
+
+        for definition in node {
+            let def_clone = definition.clone().unwrap();
```
Comment (ah-yu) [inline crates/biome_grit_formatter/src/grit/lists/definition_list.rs:11]: Is `unwrap` safe here?

File: crates/biome_grit_formatter/src/grit/lists/predicate_list.rs [hunk 4]
```diff
@@ -5,6 +5,16 @@ pub(crate) struct FormatGritPredicateList;
 impl FormatRule<GritPredicateList> for FormatGritPredicateList {
     type Context = GritFormatContext;
     fn fmt(&self, node: &GritPredicateList, f: &mut GritFormatter) -> FormatResult<()> {
-        format_verbatim_node(node.syntax()).fmt(f)
+        let mut join = f.join_nodes_with_hardline();
+
+        for predicate in node {
+            let pred_clone = predicate.clone().unwrap();
```
Comment (ah-yu) [inline crates/biome_grit_formatter/src/grit/lists/predicate_list.rs:11]: Ditto

File: crates/biome_grit_formatter/tests/spec_tests.rs [hunk 5]
```diff
@@ -3,6 +3,10 @@ mod spec_test;
 mod formatter {
 
     mod grit_module {
-        tests_macros::gen_tests! {"tests/specs/grit/**/*.grit", crate::spec_test::run, ""}
+        tests_macros::gen_tests! {"tests/specs/grit/*.grit", crate::spec_test::run, ""}
```
Comment (ah-yu) [inline crates/biome_grit_formatter/tests/spec_tests.rs:6]: What is the intention of this file change?

File: crates/biome_grit_formatter/tests/specs/grit/options.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (ah-yu) [inline crates/biome_grit_formatter/tests/specs/grit/options.json:1]: Wonder if this file can be deleted since it contains no content

File: crates/biome_grit_formatter/tests/specs/grit/patterns/options.json [hunk 7]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (ah-yu) [inline crates/biome_grit_formatter/tests/specs/grit/patterns/options.json:1]: Ditto

File: crates/biome_grit_formatter/tests/spec_tests.rs (hunk 5 above)
Comment (branberry) [reply crates/biome_grit_formatter/tests/spec_tests.rs:6]: Ah, I broke it a part to better debug the individual tests. I was using the VSCode extension to run the debugger, and for whatever reason it wouldn't show the individual test that's specific to the `where` pattern`. I can move these back together though instead of being separate modules.

File: crates/biome_grit_formatter/tests/specs/grit/options.json (hunk 6 above)
Comment (branberry) [reply crates/biome_grit_formatter/tests/specs/grit/options.json]: Thanks for calling this out! I threw this in there because the formatter options weren't being read in. I thought I had already added all of the needed configuration for the defaults, but I was missing a few extra functions which I added in for this PR.

File: crates/biome_grit_formatter/src/grit/lists/definition_list.rs (hunk 3 above)
Comment (branberry) [reply crates/biome_grit_formatter/src/grit/lists/definition_list.rs:11]: Thanks for calling these out as well! I added these in earlier to help with debugging, but it's not necessary to clone here.

Comment (branberry) [reply crates/biome_grit_formatter/src/grit/lists/definition_list.rs:11]: Thanks for calling this out! I was following an example from the GraphQL formatter for [definition lists](https://github.com/branberry/biome/blob/a6593fa5e1e524bc2859fb6916cfd3404dba40c6/crates/biome_graphql_formatter/src/graphql/lists/definition_list.rs#L6), but I didn't fully understand why I was getting a `Result<AnyGritDefinition, SyntaxError>` back as opposed to just `AnyGritDefinition`, but I realize now it's due to [the GritDefinitionList implementing the AstSeparatedList trait](https://github.com/branberry/biome/blob/a6593fa5e1e524bc2859fb6916cfd3404dba40c6/crates/biome_grit_syntax/src/generated/nodes.rs#L12625) as opposed to [GraphQL implementing the AnyNodeList trait](https://github.com/branberry/biome/blob/a6593fa5e1e524bc2859fb6916cfd3404dba40c6/crates/biome_graphql_syntax/src/generated/nodes.rs#L7650). 

I can handle the the error here and not use `unwrap`, but I'm now wondering if the `AstSeparatedList` makes sense for the `AnyGritDefinition` struct. @arendjr, any thoughts on this?

Comment (arendjr) [reply crates/biome_grit_formatter/src/grit/lists/definition_list.rs:11]: I’m not sure either, to be honest. I could imagine it has to do with the newlines that must be used as separate between Grit definitions?

Comment (branberry) [reply crates/biome_grit_formatter/src/grit/lists/definition_list.rs:11]: Hey Arend! After a bit more research, I believe it might have to do with the fact that definitions also include any pattern or predicate. One example is the `or` pattern, which separates arguments by a comma. I couldn't find an instance, though, where the definition list was greater than one. I could do the work of having a separator implementation, but it might be a bit out of scope for this PR as it would probably introduce a good amount of code.

I think for now, I'll just handle the error in a more rudimentary way until we have a better idea of when we might want the separator logic.

File: crates/biome_grit_formatter/tests/spec_tests.rs (hunk 5 above)
Comment (ah-yu) [reply crates/biome_grit_formatter/tests/spec_tests.rs:6]: Thank you for explaining the reason behind this change, but I suggest reverting the changes to this file because we need to ensure that all tests in the `tests/specs/grit` folder are covered. However, after adding these changes, some tests will not be covered. For example, if we add a new test in `tests/specs/grit/predicate/foo.grit` in the future, this test file will not be covered.

File: crates/biome_grit_formatter/src/grit/lists/definition_list.rs [hunk 8]
```diff
@@ -5,6 +5,22 @@ pub(crate) struct FormatGritDefinitionList;
 impl FormatRule<GritDefinitionList> for FormatGritDefinitionList {
     type Context = GritFormatContext;
     fn fmt(&self, node: &GritDefinitionList, f: &mut GritFormatter) -> FormatResult<()> {
-        format_verbatim_node(node.syntax()).fmt(f)
+        let mut join: JoinNodesBuilder<'_, '_, Line, GritFormatContext> =
```
Comment (ematipico) [inline crates/biome_grit_formatter/src/grit/lists/definition_list.rs:8]:
```suggestion
        let mut join =
```
Usually, types are added only when the compiler can't infer them

File: crates/biome_grit_formatter/src/grit/lists/definition_list.rs [hunk 9]
```diff
@@ -5,6 +5,22 @@ pub(crate) struct FormatGritDefinitionList;
...
+            match definition {
+                Ok(definition) => {
+                    join.entry(
+                        definition.syntax(),
+                        &format_or_verbatim(definition.format()),
+                    );
+                }
+                Err(_) => (),
+            }
+        }
```
Comment (ematipico) [inline crates/biome_grit_formatter/src/grit/lists/definition_list.rs:22]: No need for using the `match`. You can simply do 

```rs
let definition = definition?;
```

File: crates/biome_service/src/file_handlers/grit.rs [hunk 10]
```diff
@@ -16,8 +16,32 @@ use super::{
...
 };
 
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+pub struct GritFormatterSettings {
+    pub line_ending: Option<LineEnding>,
+    pub line_width: Option<LineWidth>,
+    pub indent_width: Option<IndentWidth>,
+    pub indent_style: Option<IndentStyle>,
+    pub quote_style: Option<QuoteStyle>,
```
Comment (ematipico) [inline crates/biome_service/src/file_handlers/grit.rs:26]: Does grit use different quotes?

File: crates/biome_grit_formatter/src/grit/lists/definition_list.rs (hunk 3 above)
Comment (arendjr) [reply crates/biome_grit_formatter/src/grit/lists/definition_list.rs]: Sounds good to me!

File: crates/biome_service/src/file_handlers/grit.rs (hunk 10 above)
Comment (arendjr) [reply crates/biome_service/src/file_handlers/grit.rs:26]: Good call, we shouldn't need this option for Grit.

File: crates/biome_grit_formatter/tests/spec_tests.rs (hunk 5 above)
Comment (branberry) [reply crates/biome_grit_formatter/tests/spec_tests.rs]: Makes sense! Reverting.

File: crates/biome_grit_formatter/src/grit/lists/definition_list.rs (hunk 8 above)
Comment (branberry) [reply crates/biome_grit_formatter/src/grit/lists/definition_list.rs]: My VSCode is being stubborn, and keeps putting it in there whenever I delete it for some reason 😅

File: crates/biome_grit_formatter/src/grit/lists/definition_list.rs (hunk 9 above)
Comment (branberry) [reply crates/biome_grit_formatter/src/grit/lists/definition_list.rs]: Thank you for this! Rust newbie here, so I appreciate this kind of input.

**Review by ah-yu — APPROVED** [review]

**Review by arendjr — APPROVED** [review]
